### PR TITLE
Add `createUserTag` mutation [Fixes part of #1126]

### PR DESCRIPTION
**Pull request by EshaanAgg**
**What kind of change does this PR introduce?**
Feature

**Issue Number:**

Fixes part of #1126 

**Did you add tests for your changes?**
Yes

**Snapshots/Videos:**
NA

**If relevant, did you update the documentation?**
NA

**Summary**
Adds the source code and tests for `createUserTag` mutation.

**Does this PR introduce a breaking change?**
No

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**
Yes

### CODE REVIEW

**Comment by xoldyckk** [comment]
@EshaanAgg 
Take a look:- https://thepalisadoes-dyb6419.slack.com/archives/CSWH4PN0M/p1679053124253489

Are you sure implementing these new resolvers with the older approach a good idea?

**Comment by EshaanAgg** [comment]
Strong advocate for the approach you are suggesting here as the translation thing is completely broken, and unmaintainable in the long run. It should be the duty of the client to provide the translations of all the errors that are being generated. 
The same is true for the way we are handling errors to. Custom error objects are great, but they do not provide granular control over which field exactly is erroring out, and how should the client instruct the user to resolve them.

But at the same time I think, implementing either of these changes at this point should be avoided. It is simply due to consistency. If we are to devise a new strategy to handle errors and translation, the same should be done first with the approval of the majority of the community. After the same is done, we should write some at least basic migration guides from the old approach to the new approach, so that the same can be utilised by all the contributors to completely migrate the codebase.

Migration to new strategy for some newly added features, while not immediately acting on the older ones is a bad practice in my opinion. However this doesn't mean that we shouldn't be actively discussing the newer solutions and their merits, as after all, things are indeed broken right now on the API side.

**Comment by palisadoes** [comment]
> @xoldyckk Strong advocate for the approach you are suggesting here as the translation thing is completely broken, and unmaintainable in the long run. It should be the duty of the client to provide the translations of all the errors that are being generated. The same is true for the way we are handling errors to. Custom error objects are great, but they do not provide granular control over which field exactly is erroring out, and how should the client instruct the user to resolve them.
> 
> But at the same time I think, implementing either of these changes at this point should be avoided. It is simply due to consistency. If we are to devise a new strategy to handle errors and translation, the same should be done first with the approval of the majority of the community. After the same is done, we should write some at least basic migration guides from the old approach to the new approach, so that the same can be utilised by all the contributors to completely migrate the codebase.
> 
> Migration to new strategy for some newly added features, while not immediately acting on the older ones is a bad practice in my opinion. However this doesn't mean that we shouldn't be actively discussing the newer solutions and their merits, as after all, things are indeed broken right now on the API side.

@xoldyckk Can you write a brief and definitive summary of how we should handle errors in the API so that we can add it to Talawa-Docs as a reference for future developers? That way it won't get lost. Send me the markdown and I'll add it to that repo.

**Review by xoldyckk — COMMENTED** [review]
@EshaanAgg Add the logic so that two tags with same name cannot exist under the same `parentTagId`

File: src/resolvers/Mutation/createUserTag.ts [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+      requestContext.translate(TAG_ALREADY_EXISTS.MESSAGE),
+      TAG_ALREADY_EXISTS.CODE,
+      TAG_ALREADY_EXISTS.PARAM
+    );
+  }
+
+  // Creates new tag and returns the same
+  return await OrganizationTagUser.create({
+    ...args.input,
+  });
```
Comment (xoldyckk) [inline src/resolvers/Mutation/createUserTag.ts:106]: return `lean()` here

File: src/resolvers/Mutation/createUserTag.ts [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+      throw new errors.NotFoundError(
+        requestContext.translate(INCORRECT_TAG_INPUT.MESSAGE),
+        INCORRECT_TAG_INPUT.CODE,
+        INCORRECT_TAG_INPUT.PARAM
+      );
+    }
+  }
+
+  // Check if the user has privileges to create the tag
+  const currentUserIsOrganizationAdmin = currentUser.adminFor.some(
```
Comment (xoldyckk) [inline src/resolvers/Mutation/createUserTag.ts:74]: This authorization check should happen right after you confirm whether the organization exists.

**Comment by EshaanAgg** [comment]
@xoldyckk Make the required changes!

File: tests/resolvers/Mutation/createUserTag.spec.ts [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+      await createUserTagResolver?.({}, args, context);
+    } catch (error: any) {
+      expect(spy).toBeCalledWith(ORGANIZATION_NOT_FOUND_ERROR.MESSAGE);
+      expect(error.message).toEqual(
+        `Translated ${ORGANIZATION_NOT_FOUND_ERROR.MESSAGE}`
+      );
+    }
+  });
+
+  it(`throws TAG_NOT_FOUND error if the parent tag doesn't exist`, async () => {
```
Comment (xoldyckk) [inline tests/resolvers/Mutation/createUserTag.spec.ts:119]: Change the message to include something along the lines `if parentTagId is provided and is not null`.

When writing test messages, think of how easy you can make for someone reading the test to understand what the test is testing for. Testing is a kind of documentation for your code, when looking at tests the intent should be clear.

File: tests/resolvers/Mutation/createUserTag.spec.ts [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+      );
+
+      await createUserTagResolver?.({}, args, context);
+    } catch (error: any) {
+      expect(spy).toBeCalledWith(TAG_NOT_FOUND.MESSAGE);
+      expect(error.message).toEqual(`Translated ${TAG_NOT_FOUND.MESSAGE}`);
+    }
+  });
+
+  it(`throws INCORRECT_TAG_INPUT error if the parent tag doesn't belong to the provided organization`, async () => {
```
Comment (xoldyckk) [inline tests/resolvers/Mutation/createUserTag.spec.ts:149]: same as above

File: tests/resolvers/Mutation/createUserTag.spec.ts (hunk 3 above)
Comment (EshaanAgg) [reply tests/resolvers/Mutation/createUserTag.spec.ts:119]: Got it.

File: tests/resolvers/Mutation/createUserTag.spec.ts [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+      await createUserTagResolver?.({}, args, context);
+    } catch (error: any) {
+      expect(spy).toBeCalledWith(ORGANIZATION_NOT_FOUND_ERROR.MESSAGE);
+      expect(error.message).toEqual(
+        `Translated ${ORGANIZATION_NOT_FOUND_ERROR.MESSAGE}`
+      );
+    }
+  });
+
+  it(`throws TAG_NOT_FOUND error if the parent tag is provided (not null) but the same doesn't exist in the organization of the tag being created`, async () => {
```
Comment (xoldyckk) [inline tests/resolvers/Mutation/createUserTag.spec.ts:119]: Make it `parentTagId is provided(not null) but no tag exists with _id === args.input.parentTagId`

Comment (EshaanAgg) [reply tests/resolvers/Mutation/createUserTag.spec.ts:119]: Got it.

**Review by xoldyckk — APPROVED** [review]

**Review by palisadoes — APPROVED** [review]

**Pull request merged**